### DISCUSSION TITLE
enumerating concept instances

### DISCUSSION THREAD

**Discussion by astako** [opencaesar/platform#17] [Q&A]
I have two additional questions related to enumeration:

1. I tried adding enumeration in a ref concept statement, but the build failed. Is the use of enumerates keyword not allowed in a reference statement?

2. What is the advantage of using a concept to enumerate concept instances? Is it helpful for reasoning and/or queries and if so, how?

-------------------------------------------------------------------------------------------------------------------------------------
I'm pasting the answer from Maged to my initial question about use of enumeration on concepts for context for additional discussions:

The case for enumerating instances of concepts is similar in principle to that of enumerating literals for scalars. Although the syntax in OML v1 for them is different (for literals it is specifying them within an `enumerated scalar`, while for instances, it is specifying them with the `enumerates` keyword on a concept), I am unifying the syntax in OML v2 (both will be a `oneOf <comma-sep-list>` keyword you can add to a `scalar` or a `concept` within their [..] brackets).
 
For instances, the idea is that you define them in a description and then enumerate them in a vocabulary:
 
description <http://description1#> as desc1 {
uses <http://vocabulary1#> as voca1
 
                ci Earth : voca1:Planet
                ci Mars : voca1:Planet
                ci Venus : voca1.Planet
}
 
vocabulary <http://vocabulary1#> as voca1 {
uses <http://description1#> as desc1
 
                concept Planet [
                                enumerates desc1:Earth, desc1:Mars, desc1:Venus
                ]
}

**Comment by melaasar** [comment]
> I tried adding enumeration in a ref concept statement, but the build failed. Is the use of enumerates keyword not allowed in a reference statement?

Yes, in OML v1 this is not allowed. This will be allowed in upcoming OML v2. However, notice that each set you assert, whether on the definition or on a ref, will be asserted as the exclusive set. So, if you have multiple such assertions, the DL reasoner will check that they are indeed the same instances; otherwise will complain about inconsistency.

> What is the advantage of using a concept to enumerate concept instances? Is it helpful for reasoning and/or queries and if so, how?

What the `enumerates` axiom give you is asserting that this set is the exclusive instances typed by that concept. So, if the DL reasoner found some other instances (by assertion or inference) that are not the same as those in the set (even if they have a difference name), it will treat this as a contradiction.
